shap/explainers/gradient.py

```diff
@@ -245,3 +267,119 @@ def run(self, out, model_inputs, X):
         if self.keras_phase_placeholder is not None:
             feed_dict[self.keras_phase_placeholder] = 0
         return self.session.run(out, feed_dict)
+
+
+class _PyTorchGradientExplainer(Explainer):
+
+    def __init__(self, model, data, batch_size=50, local_smoothing=0):
+
+        # try and import pytorch
+        global torch
+        if torch is None:
+            import torch
+            if LooseVersion(torch.__version__) < LooseVersion("0.4"):
+                warnings.warn("Your PyTorch version is older than 0.4 and not supported.")
+
+        self.model = model.eval()
+        # check if we have multiple inputs
+        self.multi_input = False
+        if type(data) == list:
+            self.multi_input = True
+        if type(data) != list:
+            data = [data]
+
+        self.data = data
+        self.batch_size = batch_size
+        self.local_smoothing = local_smoothing
+
+        # this is wrong; multi output is multi class output
+        multi_output = False
+        outputs = self.model(*self.data)
+        if outputs.shape[1] > 1:
+            multi_output = True
+        self.multi_output = multi_output
+
+        if not self.multi_output:
+            self.gradients = [None]
+        else:
+            self.gradients = [None for i in range(outputs.shape[1])]
+
+    def gradient(self, idx, inputs):
+        self.model.zero_grad()
+        X = [x.requires_grad_() for x in inputs]
+        outputs = self.model(*X)
+        selected = [val for val in outputs[:, idx]]
+        grads = [torch.autograd.grad(selected, x_batch)[0] for x_batch in X]
+        return grads
+
+    def shap_values(self, X, nsamples=200, ranked_outputs=None, output_rank_order="max"):
+        # check if we have multiple inputs
+        if not self.multi_input:
+            assert type(X) != list, "Expected a single tensor model input!"
+            X = [X]
+        else:
+            assert type(X) == list, "Expected a list of model inputs!"
+        # assert len(self.model_inputs) == len(X), "Number of model inputs does not match the number given!"
+
+        # rank and determine the model outputs that we will explain
+        with torch.no_grad():
+            model_output_values = self.model(*X)
```

Comment:
If we are not sorting then we could skip this forward pass. So perhaps just move it inside the if statement on the next line?